### PR TITLE
Modify do_action so default ctrl-click opens tab

### PR DESCRIPTION
**Pull request by KnownNexus**
Modify do_action to use an ID number
Use item in do_action instead of currentFile
Limit launching direct actions to one time
Use full file item instead of attr-only

Implements #16067

### CODE REVIEW

File: apps/files/js/filelist.js [hunk 1]
```diff
@@ -873,6 +867,35 @@
 			}
 		},
 
+	    do_action: function(id) {
```
Comment (juliushaertl) [inline apps/files/js/filelist.js]: Maybe openAction would be a bit more clear

File: apps/files/js/filelist.js [hunk 2]
```diff
@@ -873,6 +867,35 @@
...
+		console.log("do_action: id="+id);
+		var item = this.$fileList.children().filterAttr('data-id', '' + id);
+		var attr=item.attr;
```
Comment (juliushaertl) [inline apps/files/js/filelist.js]: no need for an additional variable here

File: apps/files/js/filelist.js [hunk 3]
```diff
@@ -873,6 +867,35 @@
...
+		window.FileActions.currentFile = item.find('td');
+		this.fileActions.currentFile = item.find('td');
+		console.log("do_action: Querying types; parent is "+item.find('td').parent()+"; parent has mime type "+item.find('td').parent().attr('data-mime'));
+		console.log("do_action: Direct mime type is "+item.attr('data-mime'));
+		var mime = this.fileActions.getCurrentMimeType();
+		var type = this.fileActions.getCurrentType();
+		var permissions = this.fileActions.getCurrentPermissions();
+		console.log("do_action: mime="+mime+", type="+type+", permissions="+permissions);
+		var action = this.fileActions.getDefault(mime,type, permissions);
+		if(action) {
```
Comment (juliushaertl) [inline apps/files/js/filelist.js]:
```suggestion
		if (action) {
```

File: apps/files/js/filelist.js [hunk 4]
```diff
@@ -873,6 +867,35 @@
...
+		    action(filename, {
+			$file: item,
```
Comment (juliushaertl) [inline apps/files/js/filelist.js]: indentation seems to be odd here

File: apps/files/js/filelist.js [hunk 5]
```diff
@@ -1463,8 +1492,9 @@
 			// linkUrl
 			if (mime === 'httpd/unix-directory') {
 				linkUrl = this.linkTo(path + '/' + name);
-			}
-			else {
+			} else if (mime === 'application/vnd.oasis.opendocument.text') {
```
Comment (juliushaertl) [inline apps/files/js/filelist.js]: No need to limit this to the mime type of documents. This should work for all actions on files, just not for directories.

**Review by juliusknorr — CHANGES_REQUESTED** [review]
Nice, seems to work quite well. I've commented inline, but here are some general minor issues with the code style:

- Indentation should use tabs not spaces
- Variable names should be camel case triedActionOnce
- Please remove the console.log statements (they seem to be leftovers from debugging)

File: apps/files/js/filelist.js (hunk 5 above)
Comment (KnownNexus) [reply apps/files/js/filelist.js]: To confirm, do you want me to just replace 
`linkUrl = this.getDownloadUrl(name, path, type === 'dir');`
with
`linkUrl = this.linkTo(path) + "&editfile="+fileData.id;`
?
￼

**Comment by KnownNexus** [comment]
I'm not sure what is causing the CI to fail, could someone help me figure it out?
Also i don't believe i have permission to restart the tests, is it possible restarting them could allow them to pass?

**Comment by davepage-mcr** [comment]
Hi @juliushaertl can you please re-run the CI for this MR or help @KnownNexus understand why it is failing?

**Comment by kesselb** [comment]
> can you please re-run the CI for this MR

Just rebase this branch to get the latest changes from master and run ci again

File: apps/files/js/filelist.js [hunk 6]
```diff
@@ -405,7 +405,7 @@
 					this.setupUploadEvents(this._uploader);
 				}
 			}
-
+		    this.triediActionOnce = false;
```
Comment (kesselb) [inline apps/files/js/filelist.js]:
```suggestion
		    this.triedActionOnce = false;
```

Comment (kesselb) [reply apps/files/js/filelist.js]: \+ Indentation looks off

File: apps/files/js/filelist.js [hunk 7]
```diff
@@ -1211,7 +1226,8 @@
 		 * @param animate true to animate the new elements
 		 * @return array of DOM elements of the newly added files
 		 */
-		_nextPage: function(animate) {
+	    _nextPage: function(animate) {
+		console.log("nextPage...");
```
Comment (kesselb) [inline apps/files/js/filelist.js]:
```suggestion

```

File: apps/files/js/filelist.js [hunk 8]
```diff
@@ -1211,7 +1226,8 @@
 		 * @param animate true to animate the new elements
 		 * @return array of DOM elements of the newly added files
 		 */
-		_nextPage: function(animate) {
+	    _nextPage: function(animate) {
```
Comment (kesselb) [inline apps/files/js/filelist.js]: Indentation

File: apps/files/js/filelist.js [hunk 9]
```diff
@@ -1262,7 +1278,12 @@
 					}
 				}, 0);
 			}
-
+		if(!this.triedActionOnce) {
+		    var urlParams = OC.Util.History.parseUrlQuery();
+		    console.log("Launching editor, params="+urlParams.editfile);
```
Comment (kesselb) [inline apps/files/js/filelist.js]:
```suggestion

```

File: apps/files/js/filelist.js [hunk 10]
```diff
@@ -873,6 +864,30 @@
...
+		if (action) {
+			action(filename, {
+				$file: item,
+				fileList: this,
+				fileActions: this.fileActions,
+				dir: item.attr('data-path') || this.getCurrentDirectory()
+		    	});
+			return true;
+		} else {
+		    console.log("openAction: no action found");
```
Comment (kesselb) [inline apps/files/js/filelist.js]: Drop console.log + else

File: apps/files/js/filelist.js [hunk 11]
```diff
@@ -873,6 +864,30 @@
 			}
 		},
 
+	    openAction: function(id) {
```
Comment (kesselb) [inline apps/files/js/filelist.js]: Please check indention for this method. 
image

File: apps/files/js/filelist.js [hunk 12]
```diff
@@ -1262,7 +1278,12 @@
 					}
 				}, 0);
 			}
-
+		if(!this.triedActionOnce) {
```
Comment (kesselb) [inline apps/files/js/filelist.js]: image

Please check the indentation here

**Comment by kesselb** [comment]
https://github.com/nextcloud/server/blob/master/apps/files/tests/js/favoritesfilelistspec.js

This test is failing. CI uses https://github.com/nextcloud/server/blob/master/autotest-js.sh to run the jsunit tests. I guess you need node and some dependencies to run these locally.

**Comment by kesselb** [comment]
Hi @KnownNexus, sorry for late reply :see_no_evil: Could you rebase the branch with the latest master?

Probably @juliushaertl or @skjnldsv know how to fix the tests.

**Comment by jonathanmaw** [comment]
Hi @kesselb,

I'm @KnownNexus 's coworker. He's moved on to other things so I'm looking after this for him.

I'll have a look at these changes and see if I can understand these changes and take responsibility for them.

**Comment by azul** [comment]
Hi @KnownNexus @jonathanmaw ,

Any progress on this? I have rebased your branch and fixed the indentation to only use tabs.
The branch is here:
https://github.com/azul/server/tree/open-new_tab

I'll look into the failing tests next. Please let me know if you are still working on this so we do not duplicate efforts.

**Comment by jonathanmaw** [comment]
Hi Azul,

thanks for taking care of this. I would be very grateful for you to look into the tests, since I have lost access to the equipment I had previously used to work on this due to the coronavirus restrictions.

**Comment by juliushaertl** [comment]
Merged with #20185

**Pull request closed without merging**